web/src/main/java/org/springframework/security/web/servlet/util/matcher/PathPatternRequestMatcher.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2002-2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.web.servlet.util.matcher;
+
+import java.util.Objects;
+
+import jakarta.servlet.http.HttpServletRequest;
+
+import org.springframework.http.server.PathContainer;
+import org.springframework.http.server.RequestPath;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatchers;
+import org.springframework.web.util.ServletRequestPathUtils;
+import org.springframework.web.util.pattern.PathPattern;
+import org.springframework.web.util.pattern.PathPatternParser;
+
+/**
+ * A {@link RequestMatcher} that uses {@link PathPattern}s to match against each
+ * {@link HttpServletRequest}. The provided path should be relative to the servlet (that
+ * is, it should exclude any context or servlet path).
+ *
+ * <p>
+ * To also match the servlet, please see {@link RequestMatchers#servlet}
+ *
+ * <p>
+ * Note that the {@link org.springframework.web.servlet.HandlerMapping} that contains the
+ * related URI patterns must be using the same
+ * {@link org.springframework.web.util.pattern.PathPatternParser} configured in this
+ * class.
+ * </p>
+ *
+ * @author Josh Cummings
+ * @since 6.5
+ */
+public final class PathPatternRequestMatcher implements RequestMatcher {
+
+	private final PathPattern pattern;
+
+	/**
+	 * Creates a {@link PathPatternRequestMatcher} that uses the provided {@code pattern}.
+	 * <p>
+	 * The {@code pattern} should be relative to the servlet path
+	 * </p>
+	 * @param pattern the pattern used to match
+	 */
+	public PathPatternRequestMatcher(PathPattern pattern) {
+		this.pattern = pattern;
+	}
+
+	/**
+	 * Creates a {@link PathPatternRequestMatcher} that uses the provided {@code pattern},
+	 * parsing it with {@link PathPatternParser#defaultInstance}.
+	 * <p>
+	 * The {@code pattern} should be relative to the servlet path
+	 * </p>
+	 * @param pattern the pattern used to match
+	 */
+	public static PathPatternRequestMatcher pathPattern(String pattern) {
+		PathPatternParser parser = PathPatternParser.defaultInstance;
```

Comment:
Should we allow for using the defaultInstant? How did the user verify that this is matches the one used by Spring MVC?

Reply:
I'm not quite sure I understand the question here. I think if ppl are publishing their own `PathPatternParser`, they know that they are. As such, they should parse the URI and then call the `PathPattern` constructor.

Reply:
I think adding `PathPatternRequestMatcher#withPathPatternParser` will clarify that this is something that may need to be considered. I'll also update the migration docs to assume they need to pick the one up from MVC.

In 7, Spring Security can find MVC's PathPatternParser instance and publish a `PathPatternRequestMatcher.Builder` that people and the DSL can use.